socket_tls: add TLS 1.3 support using Mbed TLS

valeriosetti opened the pull request:
This PR adds:

1. support for TLS 1.3 from Mbed TLS;
2. a sample to showcase minimal configurations to get TLS 1.2 or TLS 1.3 TLS sockets. The sample uses native sockets (so it's buildable only on `native_sim`) because the scope here is just to test the TLS configuration of Mbed TLS. Low level details on how to get a physical connection working on some real device is out of scope of this PR. This helps in:
    - keeping the sample code minimal
    - to test it against an OpenSSL server to show interoperability.

Further details are on the readme file included in this PR.

**Question to reviewers**
This PR introduces a sample which requires an OpenSSL server in order to be tested. The sample includes a bash script to do this, so it's pretty straightforward to accomplish this task locally. Would it be possible to do the same on the CI? Because otherwise the sample cannot be tested.

valeriosetti committed:
net: sockets_tls: add support for TLS 1.3

Enables TLS 1.3 sockets based on Mbed TLS.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: net: socket_tls: test different TLS configurations with sockets

This commit adds a test to showcase how to configure TLS 1.2 and 1.3
socket connections against an OpenSSL server. Only a limited number
of combinations are available for now:

- TLS 1.2
	- RSA certificate and key exchange.
	- ECDSA certificate and ECDHE key exchange.

- TLS 1.3 only supports ephemeral (ECDHE) key exchange with/without
  session tickets.

Since the goal is to test TLS connection and not low level ethernet
functionalities or similar, the only supported platform is "native_sim"
where Linux sockets are used to connect to the OpenSSL server locally.

The idea is that the Zephyr application acts a client and tries to connect
to the OpenSSL server running on the same PC. For sake of simplificity a
bash script is provided to start the OpenSSL server properly.
For completeness a bash script is also provided to re-generate
certificates and keys.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented:
@aescolar @stephanosio albeit you are not in the "automatic reviewer list" I would like to ask you some tip on how to let the CI test the sample I'm adding in this PR. If I can make it testable then I can also change it to be a test rather than a sample...

In short the sample uses native socket to emulate a TLS connection to a server and the server is an OpenSSL. There's already a bash script included in the PR so that the sample can easily be tested locally. Do you know if there's the possibility to start the server also on the CI?

aescolar commented:
> ...There's already a bash script included in the PR so that the sample can easily be tested locally. Do you know if there's the possibility to start the server also on the CI?

I would guess so. A solution may be to write a test using twister's pytest harness https://docs.zephyrproject.org/latest/develop/test/pytest.html which would run the server and your zephyr.exe.
In you want to do a quick trial before jumping into pytest, you could create a draft PR (so you don't spam too much) and start by provisionally hijacking some other of the shell scripts in tree which are run in CI. For ex., tests/bsim/ci.net.sh (remove all its content and instead do your own thing) to try that your test runs well in CI, as you may need to install more packages or so.

tomi-font commented on samples/net/sockets/tls_configurations/README.rst:
```diff
@@ -0,0 +1,98 @@
+   * ``sample.net.sockets.tls13.psk_kex``
+      using pure PSK key exchange without any certificate.
+
+Supported platform
```
```suggestion
Supported platforms
```

tomi-font commented on samples/net/sockets/tls_configurations/README.rst:
```diff
@@ -0,0 +1,98 @@
+Supported platform
+******************
+
+Only ``native_sim`` or ``native_sim/native/64`` are supported by this sample. The
+reason is that:
```
```suggestion
Only ``native_sim`` and ``native_sim/native/64`` are supported by this sample because:
```

tomi-font commented on samples/net/sockets/tls_configurations/README.rst:
```diff
@@ -0,0 +1,98 @@
+Only ``native_sim`` or ``native_sim/native/64`` are supported by this sample. The
+reason is that:
+
+* on one hand TLS sockets are basically a software component so there is no need
```
```suggestion
* TLS sockets are basically a software component so there is no need
```

tomi-font commented on samples/net/sockets/tls_configurations/README.rst:
```diff
@@ -0,0 +1,98 @@
+How to test
+***********
+
+OpenSSL server can easily be launched through the script
+:file:`scripts/start-server.sh`. It requires 1 input parameter which is used to
```
```suggestion
The OpenSSL server can easily be launched with the
:file:`scripts/start-server.sh` script. It requires 1 input parameter which is used to
```

tomi-font commented on samples/net/sockets/tls_configurations/README.rst:
```diff
@@ -0,0 +1,98 @@
+
+OpenSSL server can easily be launched through the script
+:file:`scripts/start-server.sh`. It requires 1 input parameter which is used to
+select the proper configuration to be used:
```
Lots of properness in here. :smile:
```suggestion
select the configuration to be used:
```

tomi-font commented on samples/net/sockets/tls_configurations/prj.conf:
```diff
@@ -0,0 +1,55 @@
+# Debug log options (optional)
+# CONFIG_NET_LOG=y
+# CONFIG_NET_STATISTICS=y
+# CONFIG_MBEDTLS_LOG_LEVEL_DBG=y
```
Does this actually do anything?

valeriosetti replied:
As it is now, i.e. commented out, of course no. But since this is an example I thought it could be nice for anyone willing to perform some debug, to have all the CONFIGs requried to do so.
If you think this is wrong or confusing I will remove this

valeriosetti replied:
On a second thought I think debug messages might be of interest once I change the sample to a test. In that case I can parse that the agreed key exchange is exactly the one we were expecting.

tomi-font replied:
> As it is now, i.e. commented out, of course no.

This I knew. :smile: 

My question was on `CONFIG_MBEDTLS_LOG_LEVEL_DBG` in particular. But now with some more grepping I see it influences the logging made by the debug log function passed to Mbed TLS.

tomi-font commented on samples/net/sockets/tls_configurations/prj.conf:
```diff
@@ -0,0 +1,55 @@
+# CONFIG_MBEDTLS_LOG_LEVEL_DBG=y
+# CONFIG_MBEDTLS_DEBUG=y
+
+# Disable optimizations
```
Why?

valeriosetti replied:
Leftover from debug sessions. I can comment it out as for the debug CONFIGs or remove it (in case you prefer to have also the debug CONFIGs removed - it depends on the answer to the previous comment).

tomi-font commented on samples/net/sockets/tls_configurations/src/main.c:
```diff
@@ -0,0 +1,231 @@
+};
+#endif
+
+#define PEER_PORT 4242
```
It could be nice to have all the parameters that are also present in the `start-server.sh` script be packed together in this source file.

valeriosetti replied:
Aren't they? PSK, EC/RSA certificates, port are currently in sequence in `main.c`. Which other are missing?

tomi-font commented on samples/net/sockets/tls_configurations/src/main.c:
```diff
@@ -0,0 +1,231 @@
+ * - generated from "create-certs.sh" script
+ * - converted in C array shape in the CMakeList file
+ */
+#if defined(CONFIG_PSA_WANT_ALG_RSA_PKCS1V15_SIGN) || defined(CONFIG_PSA_WANT_ALG_RSA_PSS)
```
Why both?

valeriosetti replied:
TLS 1.2 can use both while TLS 1.3 only PSS. My idea was to make the code a bit generic so that if someone wants to play with other ciphersuites he/she does not have to do too many changes.

tomi-font commented on samples/net/sockets/tls_configurations/src/main.c:
```diff
@@ -0,0 +1,231 @@
+ * - converted in C array shape in the CMakeList file
+ */
+#if defined(CONFIG_PSA_WANT_ALG_RSA_PKCS1V15_SIGN) || defined(CONFIG_PSA_WANT_ALG_RSA_PSS)
+#define SOME_CERTIFICATE_SUPPORTED
```
Maybe rather something like
```suggestion
#define USE_CERTIFICATE
```
?

tomi-font commented on samples/net/sockets/tls_configurations/src/main.c:
```diff
@@ -0,0 +1,231 @@
+#define INVALID_SOCKET (-1)
+
+/* Keep the new line because openssl uses that to start processing the incoming data */
+#define TEST_STRING		"hello world\n"
```
```suggestion
#define TEST_STRING "hello world\n"
```

tomi-font commented on samples/net/sockets/tls_configurations/src/main.c:
```diff
@@ -0,0 +1,231 @@
+
+static int socket_fd = INVALID_SOCKET;
+static struct pollfd fds[1 + 1];
+static int nfds = 2;
```
```suggestion
```

tomi-font commented on samples/net/sockets/tls_configurations/src/main.c:
```diff
@@ -0,0 +1,231 @@
+#include <zephyr/net/net_if.h>
+#include <zephyr/sys/util.h>
+
+#define CA_CERTIFICATE_TAG 1
+#define PSK_TAG 2
```
Do you actually need a different tag for certificate and PSK? They're not both used in a same build?

valeriosetti replied:
not sure, I need to check what  TLS 1.3's psk_ephemeral and also some TLS 1.2 do. However, would this be a problem if we keep both? Would it be better if I make an `enum` here so that we have "no offset"?

tomi-font replied:
No, this was a rather minor remark. `enum` or `#define` doesn't matter to me. I was wondering specifically about the need to use two different values, because I didn't know whether PSK and certificate are both used simultaneously in a same configuration. But they actually are?

tomi-font commented on samples/net/sockets/tls_configurations/src/main.c:
```diff
@@ -0,0 +1,231 @@
+
+	setup_credentials();
+
+	k_thread_priority_set(k_current_get(), 0);
```
Maybe add a comment to explain why this?

valeriosetti replied:
Sorry, that's another leftover. I removed it.

tomi-font commented on samples/net/sockets/tls_configurations/src/main.c:
```diff
@@ -0,0 +1,231 @@
+		LOG_DBG("Received: %s", test_buf);
+	}
+
+	LOG_INF("Test %s", memcmp(TEST_STRING, test_buf, sizeof(TEST_STRING)) == 0 ?
+		"PASSED" : "FAILED");
```
FYI: For automatic testing of this, you could use a Twister harness that looks for `Test PASSED`.

valeriosetti replied:
Done, this is part of the python script run by pytest.

tomi-font commented on samples/net/sockets/tls_configurations/src/main.c:
```diff
@@ -0,0 +1,231 @@
+	}
+
+	/* Prepare file descriptor for polling */
+	fds[0].fd = eventfd(0, 0);
+	fds[0].events = POLLIN;
```
Why this? It's not used anywhere?

tomi-font reviewed (commented):
cc @rlubos @jukkar

valeriosetti commented on modules/mbedtls/Kconfig.tls-generic:
```diff
@@ -25,7 +25,17 @@ config MBEDTLS_SSL_EXPORT_KEYS
 	bool "Support for setting the supported Application Layer Protocols"
 
-endif
+endif # MBEDTLS_TLS_VERSION_1_2
```
- `MBEDTLS_SSL_PROTO_DTLS` yes, only TLS 1.2
- `MBEDTLS_SSL_ALPN` no, it seems to be supported also in 1.3 according to the `tls13-support.md`
- `MBEDTLS_SSL_EXPORT_KEYS` apparently was removed in 3.1.0 so I will remove it from Zephyr as well

tomi-font replied:
> * `MBEDTLS_SSL_PROTO_DTLS` yes, only TLS 1.2

Is it that DTLS 1.3 gets automatically activated with `MBEDTLS_TLS_VERSION_1_3`? Or there is only DTLS 1.2? Anyhow, you should probably remove `(DTLS 1.3)` from the description of `MBEDTLS_TLS_VERSION_1_3`. Actually maybe even add `(no DTLS)` instead.
 
> * `MBEDTLS_SSL_ALPN` no, it seems to be supported also in 1.3 according to the `tls13-support.md`

Great, you can enclose it with an `if` (or add a `depends on`) `MBEDTLS_TLS_VERSION_1_2 || MBEDTLS_TLS_VERSION_1_3`.

> * `MBEDTLS_SSL_EXPORT_KEYS` apparently was removed in 3.1.0 so I will remove it from Zephyr as well

:hot_face: (Don't forget a migration guide entry for this one.)

valeriosetti replied:
> Is it that DTLS 1.3 gets automatically activated with MBEDTLS_TLS_VERSION_1_3? Or there is only DTLS 1.2? 

No, DTLS 1.3 is not supported (https://github.com/Mbed-TLS/mbedtls/blob/1ad1954e8f4416d155251ba10d7eb5b7e391a835/library/ssl_tls.c#L1316). So I need to remove it from the description. Thanks for the tip!

> Don't forget a migration guide entry for this one

Ops, you're right! Actually that's needed both for this and for the whole TLS 1.3 change I guess

valeriosetti committed:
mbedtls: remove unused option MBEDTLS_SSL_EXPORT_KEYS

According to Mbed TLS changelog this feature was set default ON
since 3.1.0 release, so the build symbol is no more available.
This commit removes it from Zephyr.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented on modules/mbedtls/Kconfig.tls-generic:
```diff
@@ -110,6 +120,15 @@ config MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED
+config MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED
+	bool "TLS 1.3 ephemeral key exchange mode"
+
+config MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_EPHEMERAL_ENABLED
```
Albeit the name is a composition of `PSK` and `EPHEMERAL`, as far as I can tell this is not the same as the previous ones. This is not enforced in `check_config.h` and this is not documented in `mbedtls_config.h` so to me this means that there's not dependency between the 3 key exchanges.

valeriosetti commented on modules/mbedtls/configs/config-tls-generic.h:
```diff
@@ -427,7 +458,7 @@
 
 #if defined(CONFIG_MBEDTLS_SERVER_NAME_INDICATION) && \
-    defined(MBEDTLS_X509_CRT_PARSE_C)
+	defined(MBEDTLS_X509_CRT_PARSE_C)
```
I think that a certain point I got some complains from the CI about some initial blank space, so I just replaced all of them. Likely this was not the problem as they were preexisting, so I will revert this.

valeriosetti commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -1056,7 +1056,7 @@ static int tls_set_psk(struct tls_context *tls,
 		       struct tls_credential *psk_id)
 {
-#if defined(MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED)
+#if defined(CONFIG_MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED)
```
Yes, I think so, indeed I'm moving this to a separate commit so it's easier to backport it.

rlubos replied:
Is this really a bug though? The symbols is defined by mbed TLS.

valeriosetti replied:
Yes, `MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED` would be for TLS 1.2, but if we want to take also TLS 1.3 into account then we should change it to `MBEDTLS_SSL_HANDSHAKE_WITH_PSK_ENABLED`.
Now the question is: Zephyr has a similar kconfig `CONFIG_MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED` which seems to mimic the official Mbed TLS one and which I extended in this PR to take TLS 1.3 PSK key exchanges into account. I'm not a huge fan of duplicating things because it requires to keep the Zephyr kconfig version updated to all the changes that might happen to the Mbed TLS one. So can I remove the kconfig and rely only on the Mbed TLS symbol?

rlubos replied:
I don't think we can remove that config, it serves as a dependency for `config MBEDTLS_PSK_MAX_LEN` at the Kconfig level.

> Yes, MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED would be for TLS 1.2

> Edit: in case we remove the kconfig, this is no more of interest for the backport to 3.7 of course

Sure, but for Zephyr 3.7 there's no TLS 1.3 support, so I'm still not convinced it's really needed to backport the change. Of course I might miss something, so if there's really a reason we could backport it, no big deal, just note that backporting changes require a corresponding bug report issue open.

valeriosetti replied:
Checking more on this I found that `MBEDTLS_PSK_MAX_LEN` is actually a user configurable symbol in Mbed TLS (defined in `mbedtls_config.h`) which has a default of 32, whereas `MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED` and `MBEDTLS_SSL_HANDSHAKE_WITH_PSK_ENABLED` are both "internal" symbols, meaning that they _cannot_ be set from the user intentionally. This means that also in Mbed TLS it can happen that a user defines `MBEDTLS_PSK_MAX_LEN` to some non default value, but he does not set any PSK key exchange. This would be the same as if I remove the `CONFIG_MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED` in Zephyr and the dependency of `CONFIG_MBEDTLS_PSK_MAX_LEN` from it. This basically is what @tomi-font proposed here:

> remove it and either do nothing for MBEDTLS_PSK_MAX_LEN (would be okay and simplest to me)

valeriosetti commented on samples/net/sockets/tls_configurations/sample.yaml:
```diff
@@ -0,0 +1,41 @@
+      - EXTRA_CONF_FILE=overlay_tls13.conf;overlay_ec.conf
+    extra_configs:
+      - CONFIG_MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED=y
+  sample.net.sockets.tls13.ec_kex:
+    extra_args:
+      - EXTRA_CONF_FILE=overlay_tls13.conf;overlay_ec.conf
+    extra_configs:
+      - CONFIG_MBEDTLS_SESSION_TICKETS=y
+      - CONFIG_MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED=y
```
`CONFIG_MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED` is for TLS 1.2. TLS 1.3 ciphersuite names do not include the key exchange. TLS 1.3 `CONFIG_MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED` allows either ECDH or FFDH key exchanges with ephemeral set (selection of EC or FF depends on which groups are proposed/accepted during the key exchange).
But I agree that I can improve the name of the test as ephemeral because it's better.

valeriosetti replied:
I also just realized that there was an error in the `no_tickets` case :facepalm:

valeriosetti commented on samples/net/sockets/tls_configurations/README.rst:
```diff
@@ -0,0 +1,98 @@
+the proper server counterpart in order to work successfully.
+
+
+Regenerate certificates and keys [optional]
+*******************************************
+
+If desired EC/RSA certificates and keys used in the sample can be re-generated
+using the :file:`scripts/create-certs.sh` scripts. However this is totally
+optional and not required to run the sample.
```
It's not strictly necessary but I thought it could be useful to show how to re-generate certificate and keys using a standard Linux tool. If you think this is not important I can remove the script and this section from the readme file.

valeriosetti commented on samples/net/sockets/tls_configurations/prj.conf:
```diff
@@ -0,0 +1,55 @@
+# Kernel options
+CONFIG_MAIN_STACK_SIZE=4096
+CONFIG_INIT_STACKS=y
```
This is a left over from the sample I started from. I will remove it

valeriosetti commented on modules/mbedtls/Kconfig.tls-generic:
According to `mbedtls_config.h` documentation this seems to be specific to TLS 1.2, so I would say the answer is no.

valeriosetti commented on samples/net/sockets/tls_configurations/prj.conf:
```diff
@@ -0,0 +1,55 @@
+CONFIG_ENTROPY_GENERATOR=y
+# These are enabled by default, but we don't want them here. They will be
+# re-enabled by the proper overlay configuration file if required.
+CONFIG_MBEDTLS_KEY_EXCHANGE_RSA_ENABLED=n
+CONFIG_MBEDTLS_TLS_VERSION_1_2=n
+CONFIG_PSA_WANT_KEY_TYPE_AES=n
```
Actually `config NET_SOCKETS_SOCKOPT_TLS` has all of these set as `imply`

tomi-font replied:
Ah okay. I guess those are sensible defaults such that TLS works "out of the box" rather than having to pick specific versions/ciphersuites. @rlubos

rlubos commented on samples/net/sockets/tls_configurations/src/main.c:
```diff
@@ -0,0 +1,231 @@
+	addr.sin_port = htons(PEER_PORT);
+	inet_pton(AF_INET, "127.0.0.1", &addr.sin_addr);
+
+	socket_fd = socket(addr.sin_family, SOCK_STREAM, IPPROTO_TLS_1_2);
```
Please extend the secure protocol list with TLS_1_3:
https://github.com/zephyrproject-rtos/zephyr/blob/cba339a6f4e2de474a99531d18755518568df0e8/include/zephyr/net/net_ip.h#L77

valeriosetti replied:
Ouch, that's a bad error. Nice catch! I will fix this immediately.

valeriosetti commented on modules/mbedtls/Kconfig.tls-generic:
```diff
@@ -93,7 +103,7 @@ config MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED
 config MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED
 	bool "ECDHE-ECDSA based ciphersuite modes"
-	depends on MBEDTLS_ECDH_C && MBEDTLS_ECDSA_C
+	depends on MBEDTLS_ECDH_C && MBEDTLS_ECDSA_C || (PSA_WANT_ALG_ECDH && PSA_WANT_ALG_ECDSA)
```
Yes, but I think we'll need to make it more complete. For example simply having ECDH and ECDSA is not enough, since you also need to specify support for at least 1 curve. I think this will require some new hidden intermediate CONFIG

rlubos commented on samples/net/sockets/tls_configurations/src/main.c:
```diff
@@ -0,0 +1,231 @@
+			goto exit;
+		}
+		/* The -1 here is because sizeof() accounts for "\0" whereas recv() does not. */
+		if (ret != (sizeof(TEST_STRING) - 1)) {
```
That's risky because stream socket doesn't guarantee you won't receive data in chunks. If you expect `sizeof(TEST_STRING) - 1` bytes of data to receive, technically you should call `recv()` in a loop, until a total of `sizeof(TEST_STRING) -1` bytes is received, and only then compare the buffer content.

valeriosetti replied:
What if I specify `MSG_WAITALL` as flag for `recv`? Shouldn't this solve the problem as well?

rlubos replied:
That's also an option, maybe even simpler, just remember to replace `sizeof(test_buf)` with `sizeof(TEST_STRING) - 1` when calling `recv()`.

rlubos reviewed (commented):
A general comment, please don't mix up changes in the files between commits, for example the commit introducing the sample should introduce the sample only, and not modify `sockets_tls.c`.

For changes in mbed TLS module use `mbedtls: ` prefixed commit, for changes in socket impelmentation/net stack `net: sockets: tls: `, for sample `samples: net: sockets: `

valeriosetti commented:
> A general comment, please don't mix up changes in the files between commits, for example the commit introducing the sample should introduce the sample only, and not modify `sockets_tls.c`.

Yes, sorry, the mixing was not intentional. Changes appeared in the wrong commits due to my rebase error during the development. I should have fixed this and I will ensure this is kept in future updates as well.

valeriosetti commented on samples/net/sockets/tls_configurations/sample.yaml:
```diff
@@ -0,0 +1,41 @@
+    - native_sim/native/64
+  # Testing requires the OpenSSL to be started, but this is currently not
+  # available on the CI so we skip it. Testing is possible locally though.
+  skip: true
```
I would like to point out that this is a sample for now, but [I would like to make a test](https://github.com/zephyrproject-rtos/zephyr/pull/77567#issuecomment-2320592981) of it. So `skip` here is temporary to avoid CI failures, but this is going to be removed once I figure out how to test it properly.

rlubos replied:
So if the ultimate goal to make a TLS testsuite of it, sholdn't this be in `tests/net` instead of `samples/net`? I wouldn't mind limiting the functionality to a single platform only then.

valeriosetti replied:
Yes, I am planning to move it to `test` once I find the proper way to test it ;)

valeriosetti commented on samples/net/sockets/tls_configurations/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+
+target_sources(app PRIVATE src/main.c)
+
+# A couple of helper functions to convert a certificate input file in PEM format
```
I saw `generate_inc_file_for_target()` in the `echo_client` sample application (the one I started from), but it was not doing what I needed here.  In that sample `generate_inc_file_for_target` is used to convert a DER file (binary) to a C-array-like format. When I tried to do the same here, I got CI complaining that I could not include binary files in Zephyr so that's why I ended up making my own version of `generate_struct_content` + `scripts/convert_pem_cert.py`.

However now that you made me think about this again I realized that perhaps the first time I tried the CI complained because the file did not have the `*.der` extension for the certificate. I will give this a try to see if I can make this work with already existing functions (I agree it would be better).

rlubos replied:
Using DER certificate format will still give you a warning, even with .der extension (which IMO is a bit dumb, but ok). My point was, that instead of converting the cert in PEM format manually, we could simply convert the certificate in to C array, and let mbed TLS deal with PEM cert (unless there are valid reasons not to do so).

valeriosetti replied:
Oh, that's indeed a nice alternative. I will give it a try

rlubos commented:
> > A general comment, please don't mix up changes in the files between commits, for example the commit introducing the sample should introduce the sample only, and not modify `sockets_tls.c`.
> 
> Yes, sorry, the mixing was not intentional. Changes appeared in the wrong commits due to my rebase error during the development. I should have fixed this and I will ensure this is kept in future updates as well.

Sure, no worries, happens all the time. Just please use `net: ` component in the commit title prefix for networking changes, so it's easy for us to track networking-related changes in the history (makes our life easier when compiling release notes). So `net: sockets: `, `samples: net: ` etc.

tomi-font commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -146,11 +146,6 @@ config NET_SOCKETS_SOCKOPT_TLS
 	bool "TCP TLS socket option support"
 	imply TLS_CREDENTIALS
 	select MBEDTLS if NET_NATIVE
-	imply MBEDTLS_TLS_VERSION_1_2 if !NET_L2_OPENTHREAD
-	imply MBEDTLS_KEY_EXCHANGE_RSA_ENABLED if !NET_L2_OPENTHREAD
-	imply MBEDTLS_CIPHER_AES_ENABLED if !NET_L2_OPENTHREAD
-	imply PSA_WANT_KEY_TYPE_AES if !NET_L2_OPENTHREAD && PSA_CRYPTO_CLIENT
-	imply PSA_WANT_ALG_CBC_NO_PADDING if !NET_L2_OPENTHREAD && PSA_CRYPTO_CLIENT
```
You're likely setting yourself up for a fun time (Kconfig-wise) by removing these defaults.

valeriosetti replied:
Yes, I reverted this back. These configs are not set to "n" in the `prj.conf` file of my test with a better comment than before. I will re-enable what needed on overlay files

tomi-font commented on west.yml:
```diff
@@ -280,7 +280,7 @@ manifest:
       path: modules/lib/gui/lvgl
     - name: mbedtls
-      revision: 2f24831ee13d399ce019c4632b0bcd440a713f7c
+      revision: pull/60/head
```
This can be removed as https://github.com/zephyrproject-rtos/mbedtls/pull/60 is now in `main`.

tomi-font commented on samples/net/sockets/tls_configurations/README.rst:
```diff
@@ -0,0 +1,98 @@
+reason is that:
+
+* on one hand TLS sockets are basically a software component so there is no need
+  to test (i.e. depend) on some specific HW;
```
Actually, reading this again, the original would have been just fine. :blush:

valeriosetti commented on samples/net/sockets/tls_configurations/README.rst:
```diff
@@ -0,0 +1,97 @@
+How to test
+***********
+
+OpenSSL server can easily be launched with the :file:`scripts/start-server.sh`
```
This part has been removed because now testing is done though twister and pytest. No customized way of testing.

valeriosetti commented on samples/net/sockets/tls_configurations/scripts/start-server.sh:
```diff
@@ -0,0 +1,44 @@
+	CIPHERS="-ciphersuites TLS_AES_128_GCM_SHA256"
+	EXTRA_ARGS="-psk_identity PSK_identity -psk 0102030405 -allow_no_dhe_kex -nocert"
+else
+	echo "Unknown option $1"
```
This script was removed as now testing is done through twister and pytest.

valeriosetti commented on samples/net/sockets/tls_configurations/src/main.c:
```diff
@@ -0,0 +1,235 @@
+	 */
+	data_len = sizeof(TEST_STRING) - 1;
+
+	for (int i = 0; i < 2; i++) {
```
Unfortunately OpenSSL does not have (or I did not find it at least) an option to just echo back what you sent it. The only option is `rev` which [echoes back the data you sent in reverted order](https://docs.openssl.org/1.1.1/man1/s_server/#options). That's why I'm doing this send/receive twice and I'm checking only at the end.

valeriosetti replied:
I guess that I need to add a comment for this because it might look strange to others as well

valeriosetti commented on samples/net/sockets/tls_configurations/src/main.c:
```diff
@@ -0,0 +1,235 @@
+		}
+		if (ret != data_len) {
+			LOG_ERR("Sent %d bytes, but received %d", data_len, ret);
+			goto exit;
```
I kept the `goto` but I had the final "PASSED"/"FAILED" being always printed.

valeriosetti commented on samples/net/sockets/tls_configurations/prj.conf:
```diff
@@ -0,0 +1,49 @@
+# Kernel options
+CONFIG_MAIN_STACK_SIZE=4096
+CONFIG_POSIX_API=y
```
I think so since I'm using POSIX calls like `socket()` (https://docs.zephyrproject.org/latest/doxygen/html/group__bsd__sockets.html#ga5693d19a0bdff45a5cb09227683d8631)

tomi-font commented on tests/net/socket/tls_configurations/testcase.yaml:
```diff
@@ -0,0 +1,51 @@
+    - net.socket
+  platform_allow:
+    - native_sim
+    - native_sim/native/64
```
```suggestion
    - native_sim/native/64
  integration_platforms:
    - native_sim
```

valeriosetti commented on tests/net/socket/tls_configurations/pytest/test_app_vs_openssl.py:
Pytest magic (which is why it took me a bit to understand and work with). It looks for the `pytest` folder, for python script whose name starts with `test_` and functions that start with `test_`.

valeriosetti commented on tests/net/socket/tls_configurations/src/main.c:
```diff
@@ -0,0 +1,252 @@
+
+int main(void)
+{
+	int ret = -EIO;
```
I just wanted to have some initial value. But it's not relevant as it is overwritten few lines below without any chance to get to the end before being properly set, so I'll remove it.

tomi-font commented on tests/net/socket/tls_configurations/prj.conf:
```diff
@@ -0,0 +1,52 @@
+
+# Disable some  Kconfigs/Kconfig options that are implied by
+# CONFIG_NET_SOCKETS_SOCKOPT_TLS. These are not wrong in general, but specific
+# to a certain case (TLS 1.2 + RSA key exchanged/certificate + AES encryption).
+# What we want here instead is to have a basic kconfig in this "prj.conf" file
```
```suggestion
# to a certain case (TLS 1.2 + RSA key exchange/certificate + AES encryption).
# What we want here instead is to have a basic configuration in this "prj.conf" file
```

tomi-font commented on tests/net/socket/tls_configurations/prj.conf:
```diff
@@ -0,0 +1,55 @@
+CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG=y
+CONFIG_ENTROPY_GENERATOR=y
+
+# Disable some kconfig that are implied from CONFIG_NET_SOCKETS_SOCKOPT_TLS.
```
Sorry I wasn't explicit enough, but here I meant it as `choose whatever you prefer between Kconfigs and Kconfig options`. :smile:

tomi-font commented on modules/openthread/CMakeLists.txt:
```diff
@@ -163,6 +163,11 @@ endforeach()
+# Since Mbed TLS 3.1.0 MBEDTLS_SSL_EXPORT_KEYS was removed as build symbol and
+# it's always assumed to be enabled. Corresponding kconfig was removed from
+# Zephyr as well, but OpenThread code still uses it, so we add it here.
+target_compile_definitions(ot-config INTERFACE -DMBEDTLS_SSL_EXPORT_KEYS)
```
Hmm... Wouldn't the proper fix be to make the OpenThread code that relies on this macro to not anymore? If you don't want to do this yourself pinging relevant people could be a good idea.

valeriosetti replied:
That was an option, but on one hand this requires opening a separate PR on the OpenThread repo and on the other one I don't know what's the preferred version of Mbed TLS to be used with OpenThread. `MBEDTLS_SSL_EXPORT_KEYS` was only removed after Mbed TLS 3.1.0, so if you build OpenThread with previous versions that build symbol is still available.
This fix looks easier and safer IMO. Wdyt?

tomi-font replied:
Ah so the code that makes use of this macro in OpenThread doesn't live in the main Zephyr repo? In any case this is fine for me. Might just be worth pinging relevant people to make them aware of this.

tomi-font commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -41,6 +41,11 @@ Mbed TLS
   have been removed because Mbed TLS doesn't support TLS 1.0 and 1.1 anymore since v3.0. (:github:`76833`)
 
+* The Kconfig option ``CONFIG_MBEDTLS_SSL_EXPORT_KEYS`` was removed because the
+  corresponding build symbol was removed from Mbed TLS since revision 3.1.0.
```
Rather either 
```suggestion
  corresponding build symbol was removed in Mbed TLS 3.1.0.
```
or
```suggestion
  corresponding build symbol doesn't exist anymore since Mbed TLS 3.1.0.
```
and up to you whether you want to add `version`.

tomi-font commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -41,6 +41,11 @@ Mbed TLS
 
+* The Kconfig option ``CONFIG_MBEDTLS_SSL_EXPORT_KEYS`` was removed because the
+  corresponding build symbol was removed from Mbed TLS since revision 3.1.0.
+  ``MBEDTLS_SSL_EXPORT_KEYS`` is assumed to be default on in Mbed TLS.
```
I think this could be rephrased to be more clear. To me this sounds like the macro may still exist in Mbed TLS, which it doesn't. Maybe something like:
```suggestion
  Mbed TLS has the functionality that ``MBEDTLS_SSL_EXPORT_KEYS`` used to enable enabled by default, so uses of ``CONFIG_MBEDTLS_SSL_EXPORT_KEYS`` can simply be removed.
```

valeriosetti replied:
Mmmm, I still prefer my version: it has the same meaning and less "enable" repetitions. But if you strongly suggest to change this, I can of course

tomi-font replied:
It's fine. I just find the `assumed to be default on` possibly a bit misleading. Maybe just changing it to `assumed to be enabled` to at least remove the `default` would be more clear and okay for you?

tomi-font commented on doc/releases/release-notes-4.0.rst:
```diff
@@ -315,6 +315,17 @@ Libraries / Subsystems
 
   * Mbed TLS was updated to version 3.6.1. The release notes can be found at:
     https://github.com/Mbed-TLS/mbedtls/releases/tag/mbedtls-3.6.1
+  * Kconfig symbol :kconfig:option:`CONFIG_MBEDTLS_TLS_VERSION_1_3` was added to
+    enable TLS 1.3 support from Mbed TLS. When this is enabled then also the
+    following new kconfig symbols can be enabled:
```
```suggestion
  * The Kconfig symbol :kconfig:option:`CONFIG_MBEDTLS_TLS_VERSION_1_3` was added to
    enable TLS 1.3 support from Mbed TLS. When this is enabled the
    following new kconfig symbols can also be enabled:
```

tomi-font commented on tests/net/socket/tls_configurations/README.rst:
```diff
@@ -0,0 +1,65 @@
+Overview
+********
+
+This sample showcases various minimal Mbed TLS configurations to properly set up
```
This is not a sample anymore. Plus, is this documentation accessible anywhere? It could even be removed completely.

valeriosetti replied:
Back to when this really was a sample it was (I checked), but I did not re-checked since I moved it to the `tests`. Once I can have the HTML build fine again, I will check again.
I think it makes sense to have a small README just to explain what this test is about. Do you think it's worth it? Or is it clear from other comments in the test?

tomi-font replied:
I don't think it's really worth it. The code speaks for itself (especially the `testcase.yaml` file), plus I don't think we have documentation for tests?

tomi-font commented on tests/net/socket/tls_configurations/prj.conf:
```diff
@@ -0,0 +1,55 @@
+# CONFIG_MBEDTLS_LOG_LEVEL_DBG=y
+# CONFIG_MBEDTLS_DEBUG=y
+
+# Disable optimizations (optional)
```
You made it disappear?

valeriosetti replied:
Yes, in the end that's something the user can enable on his own in case he needs to do any debug. Debug messages on the lines above should be enough

tomi-font commented on tests/net/socket/tls_configurations/prj.conf:
```diff
@@ -0,0 +1,52 @@
+CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG=y
+CONFIG_ENTROPY_GENERATOR=y
+
+# Disable some  Kconfigs that are implied by CONFIG_NET_SOCKETS_SOCKOPT_TLS.
```
```suggestion
# Disable some Kconfigs that are implied by CONFIG_NET_SOCKETS_SOCKOPT_TLS.
```

valeriosetti commented:
Sorry, I had to rebase due to a conflict with `main` on the `release-note-4.0`. No other change was done

tomi-font commented on doc/releases/release-notes-4.0.rst:
```diff
@@ -329,6 +329,18 @@ Libraries / Subsystems
     is also enabled. This is only meant to be used for test purposes, not in production.
     (:github:`76408`)
+  * The Kconfig symbol :kconfig:option:`CONFIG_MBEDTLS_TLS_VERSION_1_3` was added to
+    enable TLS 1.3 support from Mbed TLS. When this is enabled then also the
+    following new kconfig symbols can also be enabled:
```
```suggestion
    enable TLS 1.3 support from Mbed TLS. When this is enabled the
    following new Kconfig symbols can also be enabled:
```

tomi-font commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -51,6 +51,11 @@ Mbed TLS
   * ``CONFIG_MBEDTLS_ZEPHYR_ENTROPY`` is now :kconfig:option:``CONFIG_MBEDTLS_ENTROPY_POLL_ZEPHYR``.
 
+* The Kconfig option ``CONFIG_MBEDTLS_SSL_EXPORT_KEYS`` was removed because the
+  corresponding build symbol was removed in Mbed TLS 3.1.0. Starting from this
+  release ``MBEDTLS_SSL_EXPORT_KEYS`` is assumed to be enabled in Mbed TLS.
```
Maybe this would be cleared:
```suggestion
  corresponding build symbol was removed in Mbed TLS 3.1.0 and is now assumed to be enabled.
```
Because now reading this, I thought that `Starting from this release` could refer to `this release` of Zephyr, not the Mbed TLs one you are referring to.

tomi-font commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -51,6 +51,11 @@ Mbed TLS
 
+* The Kconfig option ``CONFIG_MBEDTLS_SSL_EXPORT_KEYS`` was removed because the
+  corresponding build symbol was removed in Mbed TLS 3.1.0 and is now assumed to
+  be enabled.
+  (:github:`77657`)
```
```suggestion
  be enabled. (:github:`77657`)
```

tomi-font replied:
Yeah it doesn't. Okay, didn't know there were other such, too bad but any way is fine for me!

tomi-font reviewed (approved):
:smile:
